File: src/shared/Components.lua
--[=[
    @class Components
    @server @client

    This is a single store of truth for any [Component](https://eryn.io/matter/api/Component)
    that is present in the world.
]=]

local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Packages = ReplicatedStorage.Packages
local Matter = require(Packages.Matter)

local components = {}

--[=[
    @within Components
    @private

    Helper function to create a component in Matter.
]=]
local function createComponent(name: string, defaultData: any)
    components[name] = Matter.component(name, defaultData)
end

-------------------------------------------------------------------------------
-- Component definitions
-------------------------------------------------------------------------------

createComponent("Animation", {})
export type AnimationComponent = {}

createComponent("Health", {})
export type HealthComponent = {
    currentHealth: number,
    initialHealth: number,
    maxHealth: number,
}

createComponent("Model", {})
export type ModelComponent = {
    model: Model,
}

createComponent("Movement", {})
export type MovementComponent = {
    movementSpeed: number,
}

createComponent("Selected", {
    selected = false,
})
export type SelectedComponent = {}

createComponent("Selectable", {})
export type SelectableComponent = {}

createComponent("Target", {})
export type TargetComponent = {
    location: Vector3,
}

createComponent("Transform", {})
export type TransformComponent = {
    position: Vector3,
    orientation: number,
}

createComponent("Unit", {})
export type UnitComponent = {
    appearance: Model,
    owner: Player,
}

return components

File: src/shared/components.ts
import { component } from "@rbxts/matter";

export const Health = component<{
	currentHealth: number;
	initialHealth: number;
	maxHealth: number;
}>("Health");
export type Health = ReturnType<typeof Health>;

/**
 * Represents an entity that can move.
 *
 * @param movementSpeed grids per second
 * @param maxForce rate of acceleration
 */
export const Movement = component<{
	movementSpeed: number;
	maxForce: number;
}>("Movement");
export type Movement = ReturnType<typeof Movement>;

export const Renderable = component<{
	model: Model;
}>("Renderable");
export type Renderable = ReturnType<typeof Renderable>;

export const Selected = component<{
	selected: boolean;
}>("Selected", {
	selected: false,
});
export type Selected = ReturnType<typeof Selected>;

export const Selectable = component<{}>("Selectable");
export type Selectable = ReturnType<typeof Selectable>;

export const Target = component<{
	location: Vector3;
}>("Target");
export type Target = ReturnType<typeof Target>;

export const Transform = component<{
	orientation: number;
	position: Vector3;
}>("Transform");
export type Transform = ReturnType<typeof Transform>;

export const Unit = component<{
	appearance: Model;
	owner: Player;
	team?: Team;
}>("Unit");
export type Unit = ReturnType<typeof Unit>;
